File: backend/domain/key.go
package domain

import (
	"context"
	"fmt"
	"time"
)

type ID int64

type Budget struct {
	Limit  float64
	Usage  float64
	Period string // "monthly", "weekly", "none"
}

type RateLimit struct {
	Limit  int
	Period string // "second", "minute", "none"
}

type Key struct {
	ID          ID
	Name        string
	Provider    string // "openai", "anthropic", etc.
	KeyHash     string
	Prefix      string
	ExpiresAt   *time.Time
	Budget      Budget
	RateLimit   RateLimit
	IsMock      bool
	MockConfig  string
	LastResetAt time.Time
	CreatedAt   time.Time
}

func (k *Key) IsExpired() bool {
	if k.ExpiresAt == nil {
		return false
	}
	return time.Now().After(*k.ExpiresAt)
}

func (k *Key) IsBudgetExceeded() bool {
	if k.Budget.Limit <= 0 {
		return false
	}
	return k.Budget.Usage >= k.Budget.Limit
}

func (k *Key) NeedsReset() bool {
	if k.Budget.Period == "" || k.Budget.Period == "none" {
		return false
	}

	now := time.Now()
	var duration time.Duration
	switch k.Budget.Period {
	case "weekly":
		duration = 7 * 24 * time.Hour
	case "monthly":
		duration = 30 * 24 * time.Hour
	default:
		return false
	}

	return now.After(k.LastResetAt.Add(duration))
}

func (k *Key) ResetUsage() {
	k.Budget.Usage = 0
	k.LastResetAt = time.Now()
}

func (k *Key) Validate() error {
	if k.Name == "" {
		return fmt.Errorf("key name is required")
	}
	if k.Provider == "" {
		return fmt.Errorf("provider is required")
	}

	switch k.Budget.Period {
	case "monthly", "weekly", "none", "":
		// OK
	default:
		return fmt.Errorf("invalid budget period: %s", k.Budget.Period)
	}

	switch k.RateLimit.Period {
	case "second", "minute", "none", "":
		// OK
	default:
		return fmt.Errorf("invalid rate limit period: %s", k.RateLimit.Period)
	}

	return nil
}

type Repository interface {
	Save(ctx context.Context, k *Key) error
	GetByID(ctx context.Context, id ID) (*Key, error)
	GetByHash(ctx context.Context, hash string) (*Key, error)
	List(ctx context.Context) ([]*Key, error)
	Update(ctx context.Context, k *Key) error
	Delete(ctx context.Context, id ID) error
	IncrementUsage(ctx context.Context, id ID, amount float64) error
	ResetUsage(ctx context.Context, id ID, lastResetAt time.Time) error
}

File: ui/src/domain/key/Key.ts
export interface KeyProps {
    id: number;
    name: string;
    provider: string;
    key_hash: string;
    prefix: string;
    budget: {
        limit: number;
        period: string;
        usage: number;
    };
    rate_limit: {
        limit: number;
        period: string;
    };
    is_mock: boolean;
    mock_config: string;
    last_reset_at: string;
    created_at: string;
    expires_at?: string;
}

export class Key {
    constructor(private props: KeyProps) { }

    get id(): number { return this.props.id; }
    get name(): string { return this.props.name; }
    get provider(): string { return this.props.provider; }
    get keyHash(): string { return this.props.key_hash; }
    get prefix(): string { return this.props.prefix; }
    get budget(): KeyProps['budget'] { return this.props.budget; }
    get rateLimit(): KeyProps['rate_limit'] { return this.props.rate_limit; }
    get isMock(): boolean { return this.props.is_mock; }
    get mockConfig(): string { return this.props.mock_config; }
    get lastResetAt(): Date { return new Date(this.props.last_reset_at); }
    get createdAt(): Date { return new Date(this.props.created_at); }
    get expiresAt(): Date | undefined {
        return this.props.expires_at ? new Date(this.props.expires_at) : undefined;
    }

    get isActive(): boolean {
        if (!this.expiresAt) return true;
        return this.expiresAt > new Date();
    }

    get usagePercentage(): number {
        if (this.props.budget.limit === 0) return 0;
        return (this.props.budget.usage / this.props.budget.limit) * 100;
    }

    toJSON(): KeyProps {
        return { ...this.props };
    }
}
